### PR TITLE
Add support for dynamic JWS signature algorithm with JWKs - Issue 7160

### PR DESCRIPTION
**Pull request by hitchan**
CLA has been submitted.

This PR adds support for creating a JWSVerificationKeySelector dynamically based on the availability of JWK algorithms from a given JWK Set provided by an OIDC discovery document (Issue 7160). Non-standard or unknown algorithms will be silently ignored (see SignatureAlgorithm class). A warning log has also been added to let developers know if there was a problem fetching the JWK Set, although this does not halt the application from starting.

The original functionality is still in place that adds the RS256 algorithm if none are provided during initial configuration for backwards compatibility.

This functionality has been implemented for both standard and reactive versions.

### CODE REVIEW

**Comment by hitchan** [comment]
Some of the integration tests are hanging, will update the PR once the issue has been resolved.

File: oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtDecoderBuilder.java [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+	/**
+	 * Fetches {@link SignatureAlgorithm}s based on the configured {@link JWKSource}s keys.
+	 * @param jwkSource
+	 * @return A set of {@link JWSAlgorithm}s to be used for JWT signature verification.
+	 */
+	protected Set<JWSAlgorithm> getSignatureAlgorithms(JWKSource<SecurityContext> jwkSource) {
+		Set<SignatureAlgorithm> jwkAlgorithms = getDefaultAlgorithms();
+		try {
+			jwkAlgorithms.addAll(fetchSignatureVerificationAlgorithms(jwkSource));
```
Comment (jzheaux) [inline oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtDecoderBuilder.java:103]: Spring Security builders are not typically additive - instead they replace. This allows Spring Security to gracefully backoff when an application wants to manually configure a value.

What that means here is that if the application has configured any algorithms, then the auto-fetch doesn't get run.

File: oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtDecoderBuilder.java [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+	/**
+	 * Converts a list of {@link JWK}s into a set of {@link SignatureAlgorithm}s.
+	 * @param jwks
+	 * @return A set of {@link SignatureAlgorithm} instances that may be used to validate a JWT (JWS).
+	 */
+	private Set<SignatureAlgorithm> fetchSignatureVerificationAlgorithms(List<JWK> jwks) {
+		if (jwks == null) {
+			return Collections.emptySet();
+		}
+		return jwks.stream().map(jwk -> {
```
Comment (jzheaux) [inline oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtDecoderBuilder.java:172]: The Stream API is a bit too slow and hard on the garbage collector to be used in Spring Security - will you please change this to a for loop?

File: oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtDecoderBuilder.java [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+	private Set<SignatureAlgorithm> fetchSignatureVerificationAlgorithms(JWKSource<SecurityContext> jwkSource) {
+		return fetchSignatureVerificationAlgorithms(fetchSignatureVerificationJwks(jwkSource));
+	}
+
+	/**
+	 * Given a valid {@link ReactiveJWKSource}, fetches, and parses out the algorithms of available JWKs.
+	 * @param jwkSource
+	 * @return A set of {@link SignatureAlgorithm} instances that may be used to validate a JWT (JWS).
+	 */
+	private Set<SignatureAlgorithm> fetchSignatureVerificationAlgorithms(ReactiveJWKSource jwkSource) {
```
[the remark this answers is not in the document]
Comment (hitchan) [reply oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtDecoderBuilder.java:159]: Thanks for the feedback. I agree we definitely could, to be honest I don't have too much experience on the reactive side of things as of yet. This comment cleared up some doubts I had.

File: oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtDecoderBuilder.java (hunk 1 above)
Comment (hitchan) [reply oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtDecoderBuilder.java:103]: Understood, I agree that is definitely a better approach. I will make the required changes.

File: oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtDecoderBuilder.java (hunk 2 above)
Comment (hitchan) [reply oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtDecoderBuilder.java:172]: Yes that is no problem at all.

**Comment by hitchan** [comment]
@jzheaux Should I be basing these changes off of master? I am not able to get a clean build & successful test to run off of the master branch.

**Comment by jzheaux** [comment]
What I'm seeing on your branch is that the `JwtIssuerAuthenticationManagerResolver` tests are failing. This is likely because they are also using `MockWebServer`, but are only mocking the discovery endpoint at this point and not also the JWK Set endpoint.

If you added that mocking to those tests, I think you'd see those start to pass.

Also, I see that the multitenancy sample is failing, and I believe that's due to the `jwsAlgorithm` being moved from `NimbusJwtDecoderJwkSetUriBuilder` out to `JwtDecoderBuilder`. I think inlining the support you wrote back into `NimbusJwtDecoderJwkSetUriBuilder` will resolve that.

**Comment by hitchan** [comment]
@jzheaux I decided to start a new branch here https://github.com/hitchan/spring-security/tree/feature/7160-2 and inline a lot of these changes and simplify them with your comments in mind.

Could you please take a look at this branch and provide any insight as to why these (i think) simple changes cause widespread unit test failure? Most of them are "ClassDefNotFound" problems. I am unable to determine the connection between some new methods in a class and the class not being able to be found by the classloader.

What I have tried so far
 - master branch -> build -> success.
 - master branch -> Add one unused method into NimbusJwtDecoder -> 100's of unit test failures. I'm really stuck here & unsure of what to do at this point.

**Comment by hitchan** [comment]
I'm going to close this pull request in favor of the [more simplified implementation here](https://github.com/spring-projects/spring-security/pull/8752)

**Pull request closed without merging**